plugins/hls-explicit-record-fields-plugin/src/Ide/Plugin/ExplicitFields.hs

```diff
@@ -137,23 +147,42 @@ codeActionProvider ideState pId (CodeActionParams _ _ docId range _) = pluginRes
           title = "Expand record wildcard"
 
 collectRecordsRule :: Recorder (WithPriority Log) -> Rules ()
-collectRecordsRule recorder = define (cmapWithPrio LogShake recorder) $ \CollectRecords nfp -> do
-  tmr <- use TypeCheck nfp
-  let exts = getEnabledExtensions <$> tmr
-      recs = concat $ maybeToList (getRecords <$> tmr)
-  logWith recorder Debug (LogCollectedRecords recs)
-  let renderedRecs = traverse renderRecordInfo recs
-      recMap = RangeMap.fromList (realSrcSpanToRange . renderedSrcSpan) <$> renderedRecs
-  logWith recorder Debug (LogRenderedRecords (concat renderedRecs))
-  pure ([], CRR <$> recMap <*> exts)
+collectRecordsRule recorder = define (cmapWithPrio LogShake recorder) $ \CollectRecords nfp ->
+  justOrFail "Unable to TypeCheck" (use TypeCheck nfp) $ \tmr -> do
+    let exts = getEnabledExtensions tmr
+        recs = getRecords tmr
+    logWith recorder Debug (LogCollectedRecords recs)
+    let names = getNames tmr
+        renderedRecs = traverse (renderRecordInfo names) recs
+        recMap = RangeMap.fromList (realSrcSpanToRange . renderedSrcSpan) <$> renderedRecs
+    logWith recorder Debug (LogRenderedRecords (concat renderedRecs))
+    pure ([], CRR <$> recMap <*> Just exts)
+
   where
     getEnabledExtensions :: TcModuleResult -> [GhcExtension]
     getEnabledExtensions = map GhcExtension . getExtensions . tmrParsed
 
+    justOrFail :: MonadFail m => String -> m (Maybe a) -> (a -> m b) -> m b
+    justOrFail = flip . maybeM . fail
+
 getRecords :: TcModuleResult -> [RecordInfo]
 getRecords (tmrRenamed -> (hs_valds -> valBinds,_,_,_)) =
   collectRecords valBinds
 
+-- | Collects all 'Name's of a given source file, to be used
+-- in the variable usage analysis.
+getNames :: TcModuleResult -> Map UniqueKey [LocatedN Name]
+getNames (tmrRenamed -> (group,_,_,_)) = collectNames group
+
+newtype UniqueKey = UniqueKey Unique
+                  deriving newtype Eq
+
+getUniqueKey :: Name -> UniqueKey
+getUniqueKey = UniqueKey . nameUnique
+
+instance Ord UniqueKey where
+  compare (UniqueKey u1) (UniqueKey u2) = getUnique u1 `nonDetCmpUnique` getUnique u2
```

Comment:
Based on what I understood from the notes in GHC source (namely, `No Ord for Unique` and `Unique Determinism`) it should be safe to use `nonDetCmpUnique` here to be able to use `Unique`s as map keys. Still it would be good to get an approval regarding this.

Reply:
... and whatever reasoning you arrive at should go into the code as a comment

Reply:
While reading more of the GHC notes, I realized GHC provides a map type exactly for this purpose: creating maps indexed with uniques. I have opted to used that instead of rolling my own solution.